functions/go/create-setters/createsetters/create_setters.go

```diff
@@ -0,0 +1,275 @@
+checks if it is a sequence node 
+checks if all the values in the node are present to any of the ArraySetters
+adds the linecomment if they are equal 
+
+e.g. for input of Mapping node
+
+environments: 
+  - dev
+  - stage
+
+For input ApplySetters [Name: env, ValueSet: [dev, stage]], yaml node is transformed to
+
+environments: # kpt-set: ${env}
+  - dev
+  - stage
+
+e.g. for input of Mapping node with FlowStyle
+
+env: [foo, bar]
+
+For input ApplySetters [Name: env, ValueSet: [foo, bar]], yaml node is transformed to
+
+env: [foo, bar] # kpt-set: ${env}
+*/
+
+func (cs *CreateSetters) visitMapping(object *yaml.RNode, path string) error {
+	return object.VisitFields(func(node *yaml.MapNode) error {
+		if node == nil || node.Key.IsNil() || node.Value.IsNil() {
+			// don't do IsNilOrEmpty check cs empty sequences are allowed
+			return nil
+		}
+		
+		// the aim of this method is to create-setter for sequence nodes
+		if node.Value.YNode().Kind != yaml.SequenceNode {
+			// return if it is not a sequence node
+			return nil
+		}
+
+		// checks if the kind is flowstyle and adds comment to its value node 
+		// else it adds the comment to the key node 
+		changeNodeComment := node.Key
+
+		if node.Value.YNode().Style == yaml.FlowStyle {
+			changeNodeComment = node.Value
+		}
+
+		// add the key to the field path
+		fieldPath := strings.TrimPrefix(fmt.Sprintf("%s.%s", path, node.Key.YNode().Value), ".")
+
+		elements, err := node.Value.Elements()
+		if err != nil {
+			return errors.Wrap(err)
+		}
+		// extracts the values in sequence node to an array 
+		var nodeValues []string
+		for _, values := range elements {
+			nodeValues = append(nodeValues, values.YNode().Value)
+		}
+
+		for _, arraySetters := range cs.ArraySetters {
+			// checks if all the values in node are present in array setter
+			if checkEqual(nodeValues, arraySetters.ValueSet) {
+				changeNodeComment.YNode().LineComment = fmt.Sprintf("kpt-set: ${%s}", arraySetters.Name)
+				return nil
+			}
+		}
+
+		cs.Results = append(cs.Results, &Result{
+			FilePath:  cs.filePath,
+			FieldPath: fieldPath,
+			Value: changeNodeComment.YNode().LineComment,
+		})
+		return nil
+	})
+}
+
+/**
+visitScalar accepts the input scalar node and performs following steps,
+checks if it is a scalar node 
+checks if the path ends with ']' (which is the value of a sequence node)
+adds the linecomment if it's value matches with any of the setter 
+
+e.g.for input of scalar node 'nginx:1.7.1' in the yaml node
+
+apiVersion: v1
+...
+  image: nginx:1.7.1
+
+and for input CreateSetters [[name: image, value: nginx], [name: tag, value: 1.7.1]]
+The yaml node is transformed to
+
+apiVersion: v1
+...
+  image: nginx:1.7.1 # kpt-set: ${image}:${tag}
+
+*/
+
+func (cs *CreateSetters) visitScalar(object *yaml.RNode, path string) error {
+	if object.YNode().Kind != yaml.ScalarNode {
+		// return if it is not a scalar node
+		return nil
+	}
+
+	// checks if the path ends with "]"
+	if path[len(path)-1] == ']' {
```

Comment:
You don't need to check for this. You should just check for value matches. Path doesn't matter.

Reply:
This condition is to avoid the comments for the following case:
Input: 
```yaml 
roles:
  - dev
  - pro
```
Config map:
```yaml 
role: dev 
```
Wrong Output: 
```yaml 
roles:
  - dev # kpt-set: ${role}
  - pro
```
As `dev`, `pro` are the scalar nodes with path ending with roles[0], roles[1] respectively, I thought to eliminate this case, checking path would be good.

Reply:
This is correct output. You should add comment in that case.

Reply:
What should be the output for such case in flowstyle ? Should it be converted to foldedstyle with a comment ?

Reply:
Good point. So only it is better to standardize the sequence node to Folded Style. But, do it only if any of the setter value match with any of the elements in the sequence node. Then the recursive call to visitScalar will add comment as you mentioned. https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/master/functions/go/apply-setters/applysetters/apply_setters.go#L161

Reply:
Updated